javaslang/src/main/java/javaslang/collection/Vector.java

```diff
@@ -11,33 +11,82 @@
 
 import java.io.Serializable;
 import java.util.*;
+import java.util.Set;
 import java.util.function.*;
 import java.util.stream.Collector;
 
+import static javaslang.collection.Arrays2.*;
+import static javaslang.collection.Collections.*;
+import static javaslang.collection.Vector.Helper.fullBranchSize;
+import static javaslang.collection.Vector.Helper.fullTreeSize;
+import static javaslang.collection.Vector.VectorTree.emptyTree;
+
 /**
- * Vector is the default Seq implementation. It provides the best performance in between Array (with constant time element access)
- * and List (with constant time element addition).
+ * Vector is the default Seq implementation that provides effectively constant time access to any element.
+ * Many other operations (e.g. `tail`, `drop`, `slice`) are also effectively constant.
+ *
+ * It's implemented using a `bit-mapped vector trie`, i.e. a tree where each node has 32 children.
+ * For integers this means that the tree's depth will be ≤ 6.
+ *
+ * For symmetrically fast beginning and end operations (e.g. `drop` vs `dropRight`) two small arrays are used as depthless staging areas.
+ * The first and last arrays have dynamic lengths (are shifted to the correct index on drop), but the middle one has internal offset to avoid shifting.
+ * The first one is only empty if everything is empty; if the last is empty, so is the middle.
+ * The middle should only contain full blocks - padded via the offset.
  *
  * @param <T> Component type of the Vector.
- * @author Ruslan Sennov
- * @since 2.0.0
+ * @author Pap Lőrinc
+ * @since 3.0.0
  */
-public final class Vector<T> implements Kind1<Vector<?>, T>, IndexedSeq<T>, Serializable {
-
+@SuppressWarnings({"SuspiciousArrayCast", "unchecked"})
+public class Vector<T> implements Kind1<Vector<?>, T>, IndexedSeq<T> {
     private static final long serialVersionUID = 1L;
 
-    private static final Vector<?> EMPTY = new Vector<>(HashArrayMappedTrie.empty());
+    static int BRANCHING_BASE = 5;
+    static int branchingMask()   { return -1 >>> -BRANCHING_BASE; }
+    static int branchingFactor() { return 1 << BRANCHING_BASE; }
 
-    private final HashArrayMappedTrie<Integer, T> trie;
-    private final int indexShift;
+    static final Vector<?> EMPTY = new Vector<>(emptyArray(), new Vector1<>(emptyArray(), 0), emptyArray());
 
-    private Vector(HashArrayMappedTrie<Integer, T> trie) {
-        this(0, trie);
-    }
+    final VectorTree<T> middle;
+    final T[] firstLeaf, lastLeaf; // TODO we could avoid dropping these by providing a start and end offset (would be a constant memory surplus)
```

Comment:
Yes, the ones I'm using as staging areas.
> The first and last arrays have dynamic lengths (are shifted to the correct index on drop), but the middle one has internal offset to avoid shifting.
The first one is only empty if everything is empty; if the last is empty, so is the middle.
The first and last have a max size equal to the branching factor.
// TODO The middle should only contain full blocks - padded via the offset.